Sydent typing, part 3 (#1125)
matrix-org/matrix.org: gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx

- No Python 3.10 support (according to the README anyway).
- I think it might use a different kind of typing semantics; its [typing FAQ](https://google.github.io/pytype/typing_faq.html) speaks of "descriptive typing" and a more lenient approach.

#### [PyCharm](https://www.jetbrains.com/pycharm/)

PyCharm has its own means to typechecking code as you write it. It's definitely caught bugs before, and having the instant feedback as you type is really nice! I have seen it struggle with `zope.interface` and some uses of `Generic`s though.

### Runtime uses of annotations

When annotations [were first introduced](https://www.python.org/dev/peps/pep-3107/), they were a generic means to associate Python objects with parts of a program. (It was only later that the community agreed that we really want to use them to annotate _types_). These annotations are available at runtime in the `__annotations__` attribute. There's also a helper function in `typing` which will help resolve forward references.

```python
>>> from typing import get_type_hints
>>> def foo(x: int) -> str: pass
...
>>> get_type_hints(foo)
{'x': <class 'int'>, 'return': <class 'str'>}
```

> Programs and libraries are free to use these annotations at runtime as they see fit. The most well-known examples are probably [`dataclasses`](), [`attrs`](https://www.attrs.org/en/stable/) with [`auto_attribs=True`](https://www.attrs.org/en/stable/types.html) and [`Pydantic`](https://pydantic-docs.helpmanual.io/). I'd be interested to learn if anyone else is consuming annotations at runtime!

Reviewer: Did you want to add a url to the `dataclasses` link?

Author: I did, thank you!

## Summary

> All in all, in a two-week sprint we were able to get Sydent's mypy coverage from a precision of 83% up to 94%. Our work would have spotted the [bytes-versus-strings bug](https://github.com/matrix-org/sydent/pull/415); we understand why [the missing await](https://github.com/matrix-org/sydent/pull/413) wasn't detected. We [fixed](https://github.com/matrix-org/sydent/issues/419) [other](https://github.com/matrix-org/sydent/issues/449) [small](https://github.com/matrix-org/sydent/issues/447) [bugs](https://github.com/matrix-org/sydent/issues/445) too as part of the process. As well as bugs, I've hopefully made the source code clearer for future readers (but that one is hard to quantify).

Reviewer: ```suggestion
All in all, in a two-week sprint we were able to get Sydent's mypy coverage from a precision of 83% up to 94%. Our work would have spotted the [bytes-versus-strings bug](https://github.com/matrix-org/sydent/pull/415); we understand why [the missing await](https://github.com/matrix-org/sydent/pull/413) wasn't detected. We [fixed](https://github.com/matrix-org/sydent/issues/419) [other](https://github.com/matrix-org/sydent/issues/449) [small](https://github.com/matrix-org/sydent/issues/447) [bugs](https://github.com/matrix-org/sydent/issues/445) too as part of the process. As well as fix bugs, I've hopefully made the source code clearer for future readers (but that one is hard to quantify).
```

                sydent.validators             0          4            0                  0       0              0                         0
         sydent.validators.common             0         20            0                  0       0              0                         0
 sydent.validators.emailvalidator             0          8            0                  0       0              0                         0
sydent.validators.msisdnvalidator             0          8            0                  0       0              0                         0
-------------------------------------------------------------------------------------------------------------------------------------------
                            Total             9       1276          273                  0      37              0                        17

```

### The meaning of precision

There are two metrics I chose to focus on:
- the proportion of "imprecise" lines across the project; I also used the complement, `precision = 100% - imprecision`, and
- the proportion of expressions whose type is not `Any`.

> These are plotted in the graph at the top of this writeup. I could see that precision and the proportion of typed expressions were correlated, but I didn't understand how they differed. I couldn't see an explanation in the mypy docs, so I went [digging](https://github.com/python/mypy/blob/d807e097d142a88a48af93314d15ad87e41d2f19/mypy/stats.py) into the mypy source code. My understanding is as follows.

Reviewer 2: The plot is in the first post, now. Link to it?

There are plenty of articles out there about the typing. As well as the [mypy blog itself](http://mypy-lang.blogspot.com/), see Daniele Varrazzo's [post on `psycopg3` (2020)](https://www.varrazzo.com/blog/2020/03/31/first-experience-mypy/), Dropbox's [blog post (2019)](https://dropbox.tech/application/our-journey-to-type-checking-4-million-lines-of-python), Zulip's [blog post (2016)](https://blog.zulip.com/2016/10/13/static-types-in-python-oh-mypy/), Glyph's [blog post on `Protocols` (2020)](https://glyph.twistedmatrix.com/2020/07/new-duck.html) and a follow-up [comparing them to `zope.interface` (2021)](https://glyph.twistedmatrix.com/2021/03/interfaces-and-protocols.html) and Nylas's [blog (2019)](https://www.nylas.com/blog/not-your-pie-mypy/). I'm sure there's plenty more out there.

It's worth highlighting the `typeshed` project, which maintains stubs for the standard library, plus popular third-party libraries. I [submitted a PR](https://github.com/python/typeshed/pull/6150) to add a single type hint—it was a very pleasant experience! Microsoft has an [incubator](https://github.com/microsoft/python-type-stubs) of sorts for stubs too.

### Other typecheckers

> After the sprint to improve coverage, I spent a short amount of time trying the alternative type checkers out there. This is by no means an authoritative, exhaustive survey—just my quick notes.

Reviewer 2: Might be worth explicitly noting that "Mypy isn't the *only* typechecker in the Python ecosystem. Other companies have built their own tools to reason about type hints in Python, each with their own strengths and weaknesses."

Author: Thanks, d9f8bcc

- No Python 3.10 support (according to the README anyway).
- I think it might use a different kind of typing semantics; its [typing FAQ](https://google.github.io/pytype/typing_faq.html) speaks of "descriptive typing" and a more lenient approach.

#### [PyCharm](https://www.jetbrains.com/pycharm/)

PyCharm has its own means to typechecking code as you write it. It's definitely caught bugs before, and having the instant feedback as you type is really nice! I have seen it struggle with `zope.interface` and some uses of `Generic`s though.

### Runtime uses of annotations

When annotations [were first introduced](https://www.python.org/dev/peps/pep-3107/), they were a generic means to associate Python objects with parts of a program. (It was only later that the community agreed that we really want to use them to annotate _types_). These annotations are available at runtime in the `__annotations__` attribute. There's also a helper function in `typing` which will help resolve forward references.

```python
>>> from typing import get_type_hints
>>> def foo(x: int) -> str: pass
...
>>> get_type_hints(foo)
{'x': <class 'int'>, 'return': <class 'str'>}
```

> Programs and libraries are free to use these annotations at runtime as they see fit. The most well-known examples are probably ]dataclasses](https://docs.python.org/3/library/dataclasses.html), [attrs](https://www.attrs.org/en/stable/) with [auto_attribs=True](https://www.attrs.org/en/stable/types.html) and [Pydantic](https://pydantic-docs.helpmanual.io/). I'd be interested to learn if anyone else is consuming annotations at runtime!

Author: Argh. Thanks. 7779f02

## Summary

All in all, in a two-week sprint we were able to get Sydent's mypy coverage from a precision of 83% up to 94%. Our work would have spotted the [bytes-versus-strings bug](https://github.com/matrix-org/sydent/pull/415); we understand why [the missing await](https://github.com/matrix-org/sydent/pull/413) wasn't detected. We [fixed](https://github.com/matrix-org/sydent/issues/419) [other](https://github.com/matrix-org/sydent/issues/449) [small](https://github.com/matrix-org/sydent/issues/447) [bugs](https://github.com/matrix-org/sydent/issues/445) too as part of the process. As well as fix bugs, I've hopefully made the source code clearer for future readers (but that one is hard to quantify).

There's room to spin out contributions upstream too. I submitted [two](https://github.com/twisted/twisted/pull/1669) [PRs](https://github.com/twisted/twisted/pull/1668) to twisted upstream; have started to work on annotations for [pynacl](https://github.com/pyca/pynacl/pull/693) [in my](https://github.com/pyca/pynacl/pull/694) [spare time](https://github.com/pyca/pynacl/pull/695); and submitted a quick fix to [typeshed](https://github.com/python/typeshed/pull/6150).

> Looking forward, I think we'd get a quick gain from ensuring that our smaller libraries ([signedjson](https://github.com/matrix-org/python-signedjson), [canonicaljson](https://github.com/matrix-org/python-canonicaljson)) are annotated. We'll be sticking with mypy for now—`mypy-zope` is crucial given how much we make use of twisted. We've started to spent time on similar efforts to improve Sygnal and Synapse in the coming weeks—though not to the extreme standard of `--strict` across everything.

Reviewer 2: ```suggestion
Looking forward, I think we'd get a quick gain from ensuring that our smaller libraries ([signedjson](https://github.com/matrix-org/python-signedjson), [canonicaljson](https://github.com/matrix-org/python-canonicaljson)) are annotated. We'll be sticking with mypy for now—the `mypy-zope` plugin is crucial given our reliance on twisted. We're also working to improve Sygnal and Synapse—though not to the extreme standard of `--strict` across everything.
```
